dpcpp/IntelLLVM: Undefined template fmt::detail::dragonbox::float_info<long double>

With latest [dpcpp](https://www.intel.com/content/www/us/en/developer/tools/oneapi/data-parallel-c-plus-plus.html#gs.7mxi91) (Intel's LLVM/clang based SYCL compiler - which CMake classes as IntelLLVM), fmt 9.0.0 (and main branch) fails to compile with 

> `error: implicit instantiation of undefined template 'fmt::detail::dragonbox::float_info<long double>'`

Dockerfile to reproduce with current latest intel image (warning, this base image is a 20GB image/6GB download)
```dockerfile
FROM intel/oneapi-basekit:2022.2-devel-ubuntu20.04

WORKDIR /opt

RUN git clone https://github.com/fmtlib/fmt.git
RUN mkdir _build && \
    cd _build && \
    CXX=$(which dpcpp) cmake ../fmt
RUN cd _build && cmake --build .
```

Error output when running:

```
/opt/fmt/include/fmt/format.h:3098:38: error: implicit instantiation of undefined template 'fmt::detail::dragonbox::float_info<long double>'
    const auto f = basic_fp<typename info::carrier_uint>(converted_value);
                                     ^
/opt/fmt/include/fmt/format.h:3206:13: note: in instantiation of function template specialization 'fmt::detail::format_float<long double>' requested here
  int exp = format_float(convert_float(value), precision, fspecs, buffer);
            ^
/opt/fmt/include/fmt/format.h:3370:20: note: in instantiation of function template specialization 'fmt::detail::write<char, fmt::appender, long double, 0>' requested here
    return detail::write(out, value, specs, locale);
                   ^
/opt/fmt/include/fmt/core.h:1648:12: note: in instantiation of function template specialization 'fmt::detail::arg_formatter<char>::operator()<long double>' requested here
    return vis(arg.value_.long_double_value);
           ^
/opt/fmt/include/fmt/format.h:4108:26: note: in instantiation of function template specialization 'fmt::visit_format_arg<fmt::detail::arg_formatter<char> &, fmt::basic_format_context<fmt::appender, char>>' requested here
      context.advance_to(visit_format_arg(f, arg));
                         ^
/opt/fmt/include/fmt/format.h:4058:10: note: in instantiation of member function 'fmt::detail::vformat_to(buffer<char> &, basic_string_view<char>, basic_format_args<basic_format_context<detail::buffer_appender<type_identity_t<char>>, type_identity_t<char>>>, fmt::detail::locale_ref)::format_handler::on_format_specs' requested here
  struct format_handler : error_handler {
         ^
/opt/fmt/src/format.cc:34:23: note: in instantiation of function template specialization 'fmt::detail::vformat_to<char>' requested here
template FMT_API void vformat_to(buffer<char>&, string_view,
                      ^
/opt/fmt/include/fmt/format.h:1272:54: note: template is declared here
template <typename T, typename Enable = void> struct float_info;
```

As I understand it, dpcpp is "supposed" to act as a reasonably up-to-date version of clang, which depends on system standard library support (so might not have all C++20 headers e.g. `<bit>` is definitely missing, and there are also later errors with "error: no matching function for call to 'bit_cast'", but I don't know if this is caused by a cascade in the error checks so isn't the primary point of this issue).

There definitely seems to be some "long double" handling e.g. https://github.com/fmtlib/fmt/blob/81f1cc74a776581cdef8659d176049d3aeb743c6/include/fmt/core.h#L1648 so, perhaps this compiler has a slightly unusual separation of definitions?

> @akohlmey, there has been a bunch of changes to the FP logic since the issue was reported. What error do you get on the current `master`?

We just updated to version 9.1.0.  The "fmtlib_format.cpp" file is  "format.cc" from the fmt distribution, but renamed to match our compilation environment. Please find the specific error messages below. This is with the Intel OneAPI dpcpp compiler (version 2022.1.0) which does the same thing as hipcc, only for Intel GPUs. This is independent from whether I define 
`FMT_USE_LONG_DOUBLE` to 0 or 1. I can work around this with the following addition, which is clearly not correct (but makes the compiler happy).
```
  diff --git a/src/fmt/format.h b/src/fmt/format.h
  index 7c607dbd30..12670a0b87 100644
  --- a/src/fmt/format.h
  +++ b/src/fmt/format.h
  @@ -1312,6 +1312,17 @@ template <> struct float_info<double> {
     static const int shorter_interval_tie_upper_threshold = -77;
   };
   
  +#if 1
  +// LAMMPS workaround: Some LLVM based compilers (e.g. AMD's hipcc, Intel's dpcpp)
  +// fail to match the following mixed 80/128-bit template below for "long double".
  +// since they silently change long double to double. So we provide an explicit one.
  +
  +template <> struct float_info<long double> {
  +  using carrier_uint = detail::uint128_t;
  +  static const int exponent_bits = 15;
  +};
  +#endif
  +
   // An 80- or 128-bit floating point number.
   template <typename T>
   struct float_info<T, enable_if_t<std::numeric_limits<T>::digits == 64 ||
```

Here is the compilation error:
```
In file included from /home/akohlmey/compile/lammps/src/fmtlib_format.cpp:9:
In file included from /home/akohlmey/compile/lammps/src/fmt/format-inl.h:29:
/home/akohlmey/compile/lammps/src/fmt/format.h:3122:38: error: implicit instantiation of undefined template 'fmt::detail::dragonbox::float_info<long double>'
    const auto f = basic_fp<typename info::carrier_uint>(converted_value);
                                     ^
/home/akohlmey/compile/lammps/src/fmt/format.h:3230:13: note: in instantiation of function template specialization 'fmt::detail::format_float<long double>' requested here
  int exp = format_float(convert_float(value), precision, fspecs, buffer);
            ^
/home/akohlmey/compile/lammps/src/fmt/format.h:3394:20: note: in instantiation of function template specialization 'fmt::detail::write<char, fmt::appender, long double, 0>' requested here
    return detail::write(out, value, specs, locale);
                   ^
/home/akohlmey/compile/lammps/src/fmt/core.h:1655:12: note: in instantiation of function template specialization 'fmt::detail::arg_formatter<char>::operator()<long double>' requested here
    return vis(arg.value_.long_double_value);
           ^
/home/akohlmey/compile/lammps/src/fmt/format.h:4132:26: note: in instantiation of function template specialization 'fmt::visit_format_arg<fmt::detail::arg_formatter<char> &, fmt::basic_format_context<fmt::appender, char>>' requested here
      context.advance_to(visit_format_arg(f, arg));
                         ^
/home/akohlmey/compile/lammps/src/fmt/format.h:4082:10: note: in instantiation of member function 'fmt::detail::vformat_to(buffer<char> &, basic_string_view<char>, basic_format_args<basic_format_context<detail::buffer_appender<type_identity_t<char>>, type_identity_t<char>>>, fmt::detail::locale_ref)::format_handler::on_format_specs' requested here
  struct format_handler : error_handler {
         ^
/home/akohlmey/compile/lammps/src/fmtlib_format.cpp:35:23: note: in instantiation of function template specialization 'fmt::detail::vformat_to<char>' requested here
template FMT_API void vformat_to(buffer<char>&, string_view,
                      ^
/home/akohlmey/compile/lammps/src/fmt/format.h:1289:54: note: template is declared here
template <typename T, typename Enable = void> struct float_info;
                                                     ^
/home/akohlmey/compile/lammps/src/fmt/format.h:1408:46: error: implicit instantiation of undefined template 'fmt::detail::dragonbox::float_info<long double>'
    using carrier_uint = typename dragonbox::float_info<Float>::carrier_uint;
                                             ^
/home/akohlmey/compile/lammps/src/fmt/format.h:3164:42: note: in instantiation of function template specialization 'fmt::detail::basic_fp<unsigned __int128>::assign<long double, 0>' requested here
                                     : f.assign(converted_value);
                                         ^
/home/akohlmey/compile/lammps/src/fmt/format.h:3230:13: note: in instantiation of function template specialization 'fmt::detail::format_float<long double>' requested here
  int exp = format_float(convert_float(value), precision, fspecs, buffer);
            ^
/home/akohlmey/compile/lammps/src/fmt/format.h:3394:20: note: in instantiation of function template specialization 'fmt::detail::write<char, fmt::appender, long double, 0>' requested here
    return detail::write(out, value, specs, locale);
                   ^
/home/akohlmey/compile/lammps/src/fmt/core.h:1655:12: note: in instantiation of function template specialization 'fmt::detail::arg_formatter<char>::operator()<long double>' requested here
    return vis(arg.value_.long_double_value);
           ^
/home/akohlmey/compile/lammps/src/fmt/format.h:4132:26: note: in instantiation of function template specialization 'fmt::visit_format_arg<fmt::detail::arg_formatter<char> &, fmt::basic_format_context<fmt::appender, char>>' requested here
      context.advance_to(visit_format_arg(f, arg));
                         ^
/home/akohlmey/compile/lammps/src/fmt/format.h:4082:10: note: in instantiation of member function 'fmt::detail::vformat_to(buffer<char> &, basic_string_view<char>, basic_format_args<basic_format_context<detail::buffer_appender<type_identity_t<char>>, type_identity_t<char>>>, fmt::detail::locale_ref)::format_handler::on_format_specs' requested here
  struct format_handler : error_handler {
         ^
/home/akohlmey/compile/lammps/src/fmtlib_format.cpp:35:23: note: in instantiation of function template specialization 'fmt::detail::vformat_to<char>' requested here
template FMT_API void vformat_to(buffer<char>&, string_view,
                      ^
/home/akohlmey/compile/lammps/src/fmt/format.h:1289:54: note: template is declared here
template <typename T, typename Enable = void> struct float_info;
                                                     ^
/home/akohlmey/compile/lammps/src/fmt/format.h:1412:48: warning: overflow in expression; result is 2147483647 with type 'int' [-Winteger-overflow]
    const auto significand_mask = implicit_bit - 1;
                                               ^
/home/akohlmey/compile/lammps/src/fmt/format.h:1413:14: error: no matching function for call to 'bit_cast'
    auto u = bit_cast<carrier_uint>(n);
             ^~~~~~~~~~~~~~~~~~~~~~
/home/akohlmey/compile/lammps/src/fmt/format.h:297:22: note: candidate template ignored: requirement 'sizeof(int) == sizeof(long double)' was not satisfied [with To = int, From = long double]
FMT_CONSTEXPR20 auto bit_cast(const From& from) -> To {
                     ^
/home/akohlmey/compile/lammps/src/fmt/format.h:450:13: note: candidate template ignored: requirement 'sizeof(int) > sizeof(long double)' was not satisfied [with To = int, From = long double]
inline auto bit_cast(const From& from) -> To {
            ^
/home/akohlmey/compile/lammps/src/fmt/format.h:1411:47: warning: shift count >= width of type [-Wshift-count-overflow]
    const auto implicit_bit = carrier_uint(1) << num_float_significand_bits;

```

--- a/var/spack/repos/builtin/packages/fmt/package.py
+++ b/var/spack/repos/builtin/packages/fmt/package.py
@@ -13,6 +13,7 @@ class Fmt(CMakePackage):
 
     homepage = "https://fmt.dev/"
     url = "https://github.com/fmtlib/fmt/releases/download/7.1.3/fmt-7.1.3.zip"
+    maintainers = ["msimberg"]
 
     version("9.1.0", sha256="cceb4cb9366e18a5742128cb3524ce5f50e88b476f1e54737a47ffdf4df4c996")
     version("9.0.0", sha256="fc96dd2d2fdf2bded630787adba892c23cb9e35c6fd3273c136b0c57d4651ad6")

--- a/var/spack/repos/builtin/packages/fmt/package.py
+++ b/var/spack/repos/builtin/packages/fmt/package.py
@@ -70,6 +70,13 @@ class Fmt(CMakePackage):
     # Only allow [[attributes]] on C++11 and higher
     patch("fmt-attributes-cpp11_4.1.0.patch", when="@4.1.0")
 
+    # Fix compilation with hipcc/dpcpp: https://github.com/fmtlib/fmt/issues/3005
+    patch(
+        "https://github.com/fmtlib/fmt/commit/0b0f7cfbfcebd021c910078003d413354bd843e2.patch?full_index=1",
+        sha256="08fb707bf8b4fc890d6eed29217ead666558cbae38f9249e22ddb82212f0eb4a",
+        when="@9.0.0:9.1.0",
+    )
+
     def cmake_args(self):
         spec = self.spec
         args = []

